### PR TITLE
add make-sasl-ctx

### PR DESCRIPTION
**Pull request by Bogdanp**
Implements @rmculpepper 's suggested API for custom contexts from #3 .  

I ended up flipping the `out` and `next` arguments compared to your sketch to match the existing implementations, but I'm happy to change it back if you feel strongly about it.

### CODE REVIEW

File: sasl-doc/sasl.scrbl [hunk 1]
```diff
@@ -106,6 +138,24 @@ protocol is done}
...
+  sasl-next/c contract?
+  #:value (or/c 'done
+                (-> any/c
+                    (or/c bytes? string?)
+                    (values (or/c #f bytes? string?) sasl-next/c)))
+]{
+  The contract for custom SASL mechanism state transition procedures.
+}
+
+@defstruct[(exn:fail:sasl:fatal? exn:fail?)
```
Comment (rmculpepper) [inline sasl-doc/sasl.scrbl:152]: These should be the struct names, not the predicates.

File: sasl-lib/private/base.rkt [hunk 2]
```diff
@@ -1,17 +1,26 @@
...
+
+(require net/base64
+         racket/contract/base
+         racket/match)
+
 (provide (all-defined-out))
 
 ;; ----
 
+(define sasl-next/c
```
Comment (rmculpepper) [inline sasl-lib/private/base.rkt:11]: I would prefer to move this to sasl-lib/main.rkt and drop racket/contract/base require from this module.

File: sasl-doc/sasl.scrbl (hunk 1 above)
Comment (Bogdanp) [reply sasl-doc/sasl.scrbl:152]: Whoops! Will fix in a sec, thanks.

**Comment by Bogdanp** [comment]
I've amended the commit to address the two issues.

**Comment by rmculpepper** [comment]
Thanks!

**Pull request merged**